File: networkmvp/src/main/java/com/suhang/networkmvp/widget/autolayout/SoftKeyboardSizeWatchLayout.java
package com.suhang.networkmvp.widget.autolayout;

import android.app.Activity;
import android.content.Context;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.widget.RelativeLayout;

import com.suhang.networkmvp.utils.ScreenUtils;

import java.util.ArrayList;
import java.util.List;

public class SoftKeyboardSizeWatchLayout extends RelativeLayout {

	private Context mContext;
	private int mOldh = -1;
	private int mNowh = -1;
	protected int mScreenHeight = 0;
	protected boolean mIsSoftKeyboardPop = false;

	public SoftKeyboardSizeWatchLayout(final Context context, AttributeSet attrs) {
		super(context, attrs);
		this.mContext = context;
		getViewTreeObserver().addOnGlobalLayoutListener(() -> {
			Rect r = new Rect();
			((Activity) mContext).getWindow().getDecorView().getWindowVisibleDisplayFrame(r);
			if (mScreenHeight == 0) {
				mScreenHeight = r.bottom;
			}
			if (r.bottom > mScreenHeight) {
				mScreenHeight = r.bottom;
			}
			mNowh = mScreenHeight - r.bottom;
			if (mOldh != -1 && mNowh != mOldh) {
				if (mNowh > ScreenUtils.getNavigationBarHeight(context)) {
					mIsSoftKeyboardPop = true;
					if (mListenerList != null) {
						for (OnResizeListener l : mListenerList) {
							l.OnSoftPop(mNowh);
						}
					}
				} else {
					if (mListenerList != null) {
						if (mIsSoftKeyboardPop) {
							for (OnResizeListener l : mListenerList) {
								l.OnSoftClose();
							}
							mIsSoftKeyboardPop = false;
						}
					}
				}
			}
			mOldh = mNowh;
		});
	}

	public boolean isSoftKeyboardPop() {
		return mIsSoftKeyboardPop;
	}

	private List<OnResizeListener> mListenerList;

	public void addOnResizeListener(OnResizeListener l) {
		if (mListenerList == null) {
			mListenerList = new ArrayList<>();
		}
		mListenerList.add(l);
	}

	public interface OnResizeListener {
		/**
		 * 软键盘弹起
		 */
		void OnSoftPop(int height);

		/**
		 * 软键盘关闭
		 */
		void OnSoftClose();
	}
}

File: app/src/main/java/com/suhang/networkmvp/widget/autolayout/SoftKeyboardSizeWatchLayout.kt
package com.suhang.networkmvp.widget.autolayout

import android.app.Activity
import android.content.Context
import android.graphics.Rect
import android.util.AttributeSet
import android.widget.RelativeLayout


import com.suhang.networkmvp.utils.ScreenUtils

import java.util.ArrayList

open class SoftKeyboardSizeWatchLayout(private val mContext: Context, attrs: AttributeSet) : RelativeLayout(mContext, attrs) {
    private var mOldh = -1
    private var mNowh = -1
    protected var mScreenHeight = 0
    var isSoftKeyboardPop = false
        protected set

    init {
        viewTreeObserver.addOnGlobalLayoutListener {
            val r = Rect()
            (mContext as Activity).window.decorView.getWindowVisibleDisplayFrame(r)
            if (mScreenHeight == 0) {
                mScreenHeight = r.bottom
            }
            if (r.bottom > mScreenHeight) {
                mScreenHeight = r.bottom
            }
            mNowh = mScreenHeight - r.bottom
            if (mOldh != -1 && mNowh != mOldh) {
                if (mNowh > ScreenUtils.getNavigationBarHeight(mContext)) {
                    isSoftKeyboardPop = true
                    if (mListenerList != null) {
                        for (l in mListenerList!!) {
                            l.OnSoftPop(mNowh)
                        }
                    }
                } else {
                    if (mListenerList != null) {
                        if (isSoftKeyboardPop) {
                            for (l in mListenerList!!) {
                                l.OnSoftClose()
                            }
                            isSoftKeyboardPop = false
                        }
                    }
                }
            }
            mOldh = mNowh
        }
    }

    private var mListenerList: MutableList<OnResizeListener>? = null

    fun addOnResizeListener(l: OnResizeListener) {
        if (mListenerList == null) {
            mListenerList = ArrayList<OnResizeListener>()
        }
        mListenerList!!.add(l)
    }

    interface OnResizeListener {
        /**
         * 软键盘弹起
         */
        fun OnSoftPop(height: Int)

        /**
         * 软键盘关闭
         */
        fun OnSoftClose()
    }
}